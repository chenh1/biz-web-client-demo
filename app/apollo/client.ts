// import { ApolloLink, HttpLink } from "@apollo/client";
// import { onError } from "@apollo/client/link/error";

// import {
//   NextSSRInMemoryCache,
//   NextSSRApolloClient,
// } from "@apollo/experimental-nextjs-app-support/ssr";
// import { NextSSRInMemoryCache } from "@apollo/experimental-nextjs-app-support/ssr";
// import { registerApolloClient } from "@apollo/experimental-nextjs-app-support";}

// const errorLink = onError(({ graphQLErrors, networkError }) => {
//   if (graphQLErrors)
//     graphQLErrors.forEach(({ message, locations, path, ...rest }) => {
//       Object.keys(rest)?.map(key => console.log(key, rest[key]))
//       console.log(
//         `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
//       )
//     });
//   if (networkError) console.log(`[Network error]: ${networkError}`);
// });

// const httpLink = new HttpLink({
//   uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
// })

// export const { getClient } = registerApolloClient(() => {
//   return new NextSSRApolloClient({
//     cache: new NextSSRInMemoryCache(),
//     link: ApolloLink.from([errorLink, httpLink]),
//   });
// });

import { HttpLink } from "@apollo/client";
import {
  registerApolloClient,
  ApolloClient,
  InMemoryCache,
} from "@apollo/experimental-nextjs-app-support";

export const { getClient, query, PreloadQuery } = registerApolloClient(() => {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({
      // this needs to be an absolute url, as relative urls cannot be used in SSR
      uri: `${process.env.API_URL || "http://127.0.0.1:1337"}/graphql`,
      // you can disable result caching here if you want to
      // (this does not work if you are rendering your page with `export const dynamic = "force-static"`)
      // fetchOptions: { cache: "no-store" },
    }),
  });
});