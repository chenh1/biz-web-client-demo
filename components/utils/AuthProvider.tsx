'use client'

import React from 'react';
import { SessionProvider } from 'next-auth/react'

const AuthProvider: React.FC<{ children?: string | JSX.Element | JSX.Element[] | React.ReactNode }> = ({ children }) => {
    return (
        <SessionProvider>
            {children}
        </SessionProvider>
    )
}

export default AuthProvider