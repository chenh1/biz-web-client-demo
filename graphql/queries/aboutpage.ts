import { gql, DocumentNode } from "@apollo/client";

const aboutPageQuery: DocumentNode = gql`
  query AboutPage {
    aboutPage {
      data {
        attributes {
          title
          content
          topImage {
            data {
              attributes {
                url
                width
                height
                alternativeText
              }
            }
          }
        }
      }
    }
  }
`;

export default aboutPageQuery;