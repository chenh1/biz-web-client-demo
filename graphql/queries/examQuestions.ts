import { gql, DocumentNode } from "@apollo/client";

export type MultiPoints = {
  open: number;
  high: number;
  low: number;
  close: number;
}

export type ChartData = {
  multiPoints: MultiPoints;
  x: number;
  y: number;
}

export type Answer = {
  key: string;
  answer: string;
  value?: string;
  xCell?: number;
  yCell?: number;
  x?: number;
  y?: number;
}

export type Cell = {
  header: string;
  position: number;
}

export type CellValue = {
  value: string | number;
  xCell: number;
  yCell: number;
}

export type Question = {
  question: string;
  correctAnswers: Answer[];
  
}

export type InteractiveTableQuestion = Question & {
  answers: Answer[];
  tableData: CellValue[];
  rowHeaders: Cell[];
  columnHeaders: Cell[];
  answerType: string;
}

export type MultiChoiceQuestion = Question & {
  answers: Answer[];
}

export type InteractiveChartQuestion = Question & {
  xType: string;
  yType: string;
  xAxisLabel: string;
  yAxisLabel: string;
  chartType: "line" | "candlestick";
  chartData: ChartData[];
  interactiveChartData: ChartData[];
  correctAnswer: Answer;
}

export type ExamQuestionType = {
  explanation: string;
  lessonModule: number;
  multiChoiceQuestion?: MultiChoiceQuestion;
  fillInQuestion?: Question;
  interactiveTableQuestion?: InteractiveTableQuestion;
  interactiveChartQuestion?: InteractiveChartQuestion;
  tags: {
    data: {
      attributes: {
        displayName: string;
        color: string;
      }
    }
  }
}

export type ExamQuestionResponse = {
  examQuestions: {
    data: {
      attributes: ExamQuestionType
    }[]
  }
}

export type Datum = {
  [key: string]: number | string;
}

export const examQuestionsQuery: DocumentNode = gql`
  query ExamQuestions($lessonModule: Float) {
    examQuestions(filters: {lessonModule: {eq: $lessonModule }}) {
      data {
        attributes {
          tags {
            data {
              attributes {
                displayName
                color
              }
            }
          }
          explanation
          lessonModule
          multiChoiceQuestion {
            question
            answers {
              answer
              key
            }
            correctAnswers {
              key
            }
          }
          fillInQuestion {
            question
            correctAnswers {
              answer
            }
          }
          interactiveTableQuestion {
            question
            answerType
            tableData(pagination: { page: 1, pageSize: 100 }) {
              value
              xCell
              yCell
            }
            correctAnswers {
              value
              xCell
              yCell
            }
            rowHeaders {
              header
              position
            }
            columnHeaders {
              header
              position
            }
          }
          interactiveChartQuestion {
            xType
            yType
            xAxisLabel
            yAxisLabel
            question
            chartType
            chartData(pagination: { page: 1, pageSize: 100 }) {
              x
              y
              multiPoints
            }
            correctAnswer {
              x
              y
            }
          }
        }
      }
    }
  }
`