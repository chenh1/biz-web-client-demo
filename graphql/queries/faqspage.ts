import { gql, DocumentNode } from "@apollo/client";

const faqsPageQuery: DocumentNode = gql`
  query FaqsPage {
    faqsPage {
      data {
        attributes {
          title
          description
          faq(pagination: { page: 1, pageSize: 50 }) {
            question
            answer
            category {
              data {
                attributes {
                  displayName
                  value
                }
              }
            }
            answerRich
          }
        }
      }
    }
  }
`;

export default faqsPageQuery;