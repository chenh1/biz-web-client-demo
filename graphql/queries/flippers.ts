import { gql, DocumentNode } from "@apollo/client";

export const flippersQuery: DocumentNode = gql`
  query FlipperManager {
    flipperManager {
      data {
        attributes {
          ads
          adsPostsPage
          adsPostPage
          adsDashboard
          adsModuleLearnView
          adsModuleExamView
          testMode
          maintenanceMode
        }
      }
    }
  }
`;
