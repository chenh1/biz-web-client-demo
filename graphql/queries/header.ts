import { gql, DocumentNode } from "@apollo/client";

export type HeaderData = {
  navLinks: {
    url: string;
    copy: string;
  }[]
}

export type HeaderDataRes = {
  header: {
    data: {
      attributes: HeaderData
    }
  }
}

export const headerQuery: DocumentNode = gql`
  query Header {
    header {
      data {
        attributes {
          navLinks {
            url
            copy
          }
        }
      }
    }
  }
`