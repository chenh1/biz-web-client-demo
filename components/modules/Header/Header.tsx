"use client";

import type { HeaderDataRes } from "../../../graphql/queries/header";
import type { BgColor, DarkBgColor } from "../../containers/FixedDrawer/FixedDrawer";
import React, { useState } from "react"
import { FixedDrawer } from "../../containers/FixedDrawer/FixedDrawer";
import { Section } from "../../containers/Section/Section";
import { Box } from "../../containers/Box/Box";
import { Heading } from "../../core/Heading/Heading";
import { Button } from "../../core/Button/Button";
import { SpanLink } from "../../core/SpanLink/SpanLink";
import Link from "next/link"
import { usePathname, useSearchParams } from "next/navigation";
import { LoginForm } from "../LoginForm/LoginForm";
import { Overlay } from "../../containers/Overlay/Overlay";
import { useSession } from "next-auth/react";
import { signOut } from "next-auth/react";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import { Profile } from "../../modules/Profile/Profile";
import { headerQuery } from "../../../graphql/queries/header";
import { useQuery, ApolloError } from "@apollo/client";
import { useFlippers } from "../../utils/useFlippers";
import { Brand } from "../../art/Brand/Brand";

const pathsWithLightBg: string[] = [
  '/reset-password',
  '/posts'
]

interface HeaderProps {
  headerContent?: React.ReactNode;
  isTransparent?: boolean;
  bgColor?: BgColor;
  darkBgColor?: DarkBgColor;
  hasShadow?: boolean;
}

export const Header: React.FC<HeaderProps> = ({ headerContent, isTransparent, bgColor = "max-white", darkBgColor = "night-black", hasShadow = true }: HeaderProps) => {
  const pathName = usePathname();
  const searchParams = useSearchParams()
  const queryOverride = searchParams?.get('overridemaintenancemode')?.toLowerCase()
  const isOverriding = queryOverride === process.env.QUERY_OVERRIDE
  const [ showMobileNav, setShowMobileNav ] = useState<boolean>(false);
  const [ isLoggingIn, setIsLoggingIn ] = useState<boolean>(false);
  const [ isLoggingOut, setIsLoggingOut ] = useState<boolean>(false);
  const [ isAtTop, setIsAtTop ] = useState<boolean>(true);

  const session = useSession()
  const isOnPageWithLightBg = pathsWithLightBg.includes(pathName ?? '')

  const { 
    data,
    loading, 
    error 
  }: { data: HeaderDataRes | undefined; loading?: boolean; error?: ApolloError; } = useQuery(headerQuery);
  const navLinks = data?.header?.data?.attributes?.navLinks
  const { maintenanceMode } = useFlippers({ key: 'maintenanceMode' });

  return (
    <>
      <FixedDrawer 
        hasScroll 
        bgColor={isTransparent || isAtTop ? "bg-transparent" : bgColor} 
        darkBgColor={isTransparent ? "bg-transparent" : darkBgColor} 
        hasShadow={hasShadow && !isAtTop} 
        openFrom="top" 
        hasTab={false} 
        hasSensors={false} 
        onTopReached={() => setIsAtTop(true)}
        onTopLeave={() => setIsAtTop(false)}
        padding="0"
        shouldDisableBodyScroll={false}
        className={`px-0 py-1 w-screen duration-300 ease-in-out ${isAtTop && !isOnPageWithLightBg ? 'dark' : ''} ${showMobileNav ? "min-h-full md:min-h-[56px]" : "min-h-[56px]"}`}
        defaultOpenState={true}
      >
        <Section heightClass="py-1">
          <Box padding="0" className="grid-cols-2">
            {headerContent ||
              <Link href="/">
                <Brand/>
              </Link>
            }
            <div onClick={() => setShowMobileNav(!showMobileNav)} className="md:hidden cursor-pointer flex flex-col max-w-min justify-center justify-self-end">
              <div className={`${showMobileNav ? "animate-hamburgertochevronleft" : "animate-hamburgertochevronleftreverse "} border-night-black dark:border-max-white border-2 w-6`} />
              <div className={`${showMobileNav ? "animate-hamburgertochevronright" : "animate-hamburgertochevronrightreverse "} border-night-black dark:border-max-white border-2 w-6 mt-2`} />
            </div>
            <nav className={`col-span-2 md:col-span-1 duration-300 ease-in-out ${showMobileNav ? `flex h-full bg-${bgColor || 'transparent'} dark:bg-${darkBgColor || 'transparent'}`  : 'hidden h-0 md:flex md:h-full'} md:flex items-center justify-end`}>
              <ul className="flex flex-col md:flex-row items-center justify-end">
                {navLinks?.map(({ copy, url }, i) => (
                  <li key={i} className="inline mt-8 ml-12 md:mt-0">
                    <Link data-cy={`nav-link-${copy?.replace(' ', '-')?.toLowerCase()}`} href={url}>
                      <SpanLink uncolored>
                        {copy}
                      </SpanLink>
                    </Link>
                  </li>  
                ))}
                {(!maintenanceMode || isOverriding) &&
                  <li className="inline mt-8 ml-12 md:mt-0">
                    {
                      session?.status === "authenticated" ?
                        <Profile setIsLoggingOut={setIsLoggingOut}/>
                        :
                        <a data-cy="header-login-link" onClick={()=> setIsLoggingIn(!isLoggingIn)}>
                          <SpanLink uncolored>Login</SpanLink>
                        </a>
                    }
                  </li>
                }
              </ul>
            </nav>
          </Box>
        </Section>
      </FixedDrawer>
      
      {isLoggingIn &&
        <Overlay wrapperClassName="px-4 w-full sm:w-auto" visibility={isLoggingIn} toggleVisibility={setIsLoggingIn}>
          <LoginForm defaultIsRegistering={false} toggleVisibility={setIsLoggingIn}/>
        </Overlay>
      }

      {isLoggingOut &&
        <Overlay wrapperClassName="px-4 w-full sm:w-auto" visibility={isLoggingOut} toggleVisibility={setIsLoggingOut}>
          <Box padding="0" gap="md">
            <Heading type="h4" className="text-center">Logout</Heading>
            <Paragraph className="text-center">Are you sure you want to logout?</Paragraph>
            <Box padding="0" gap="md" className="grid-cols-[auto_auto]">
              <Button onClick={() => setIsLoggingOut(false)} label={"Cancel"} />
              <Button onClick={() => signOut()} label={"Logout"} primary />
            </Box>
          </Box>
        </Overlay>
      }
    </>
  )
}