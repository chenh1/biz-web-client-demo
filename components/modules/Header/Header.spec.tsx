import { Header } from "./Header";
import { MockedProvider } from "@apollo/client/testing";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
import AuthProvider from "../../utils/AuthProvider";
import { headerQuery } from "../../../graphql/queries/header";
import { flippersQuery } from "../../../graphql/queries/flippers";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock("next/navigation", () => ({
  useSearchParams() {
    return {
      get: () => 'test'
    }
  },
  usePathname() {
    return "/test"
  }
}));

jest.mock("react-markdown", () => ({ children }) => <div>{children}</div>);
jest.mock("rehype-raw", () => ({ children }) => <div>{children}</div>);
jest.mock("next/link", () => ({ url, children }) => <a href={url}>children</a>);
jest.mock('next-auth/react', () => ({
  useSession: jest.fn()
}))
import { useSession } from "next-auth/react";
const mockedUseSession = useSession as jest.Mock;

const mocks = [
  {
    request: {
      query: headerQuery
    },
    result: {
      data: {
        header: {
          data: {
            attributes: {
              navLinks: [
                {
                  url: "https://test.com",
                  copy: "Test Link"
                },
                {
                  url: "https://test2.com",
                  copy: "Test Link 2"
                }
              ]
            }
          }
        }
      }
    }
  },
  {
    request: {
      query: flippersQuery
    },
    result: {
      data: {
        flipperManager: {
          data: {
            attributes: { maintenanceMode: false }
          }
        }
      }
    }
  }
];


it('Header module with profile icon if logged in', async () => {
  mockedUseSession.mockImplementation(() => ({ status: "authenticated" }));

  const { container } = render(
    <MockedProvider mocks={mocks}>
      <Header />
    </MockedProvider>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(container).toMatchSnapshot();
});


it('Header module with login link if logged out', async () => {
  mockedUseSession.mockImplementation(() => ({ status: "unauthenticated" }));

  const { container } = render(
    <MockedProvider mocks={mocks}>
      <Header />
    </MockedProvider>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(container).toMatchSnapshot();
});
