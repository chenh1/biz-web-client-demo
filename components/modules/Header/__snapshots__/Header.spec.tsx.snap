// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header module with login link if logged out 1`] = `
<div>
  <div
    class="z-50 fixed duration-300 ease-in-out top-0  p-2  dark:bg-night-black-300 px-0 py-1 w-screen duration-300 ease-in-out dark min-h-[56px] translate-y-0"
  >
    <div
      class=" flex border-box justify-around w-full py-1 "
    >
      <div
        class="w-full max-w-7xl px-4 xl:px-0"
      >
        <div
          class=" grid  gap-4  p-2 grid-cols-2"
        >
          <a>
            children
          </a>
          <div
            class="md:hidden cursor-pointer flex flex-col max-w-min justify-center justify-self-end"
          >
            <div
              class="animate-hamburgertochevronleftreverse  border-night-black dark:border-max-white border-2 w-6"
            />
            <div
              class="animate-hamburgertochevronrightreverse  border-night-black dark:border-max-white border-2 w-6 mt-2"
            />
          </div>
          <nav
            class="col-span-2 md:col-span-1 duration-300 ease-in-out hidden h-0 md:flex md:h-full md:flex items-center justify-end"
          >
            <ul
              class="flex flex-col md:flex-row items-center justify-end"
            >
              <li
                class="inline mt-8 ml-12 md:mt-0"
              >
                <a>
                  children
                </a>
              </li>
              <li
                class="inline mt-8 ml-12 md:mt-0"
              >
                <a>
                  children
                </a>
              </li>
              <li
                class="inline mt-8 ml-12 md:mt-0"
              >
                <a
                  data-cy="header-login-link"
                >
                  <span
                    class="  relative cursor-pointer font-body text-night-black dark:text-max-white before:content-[""] before:absolute before:translate-y-[6px] before:bottom-0 before:duration-300 before:ease-in-out before:border-apex-blue before:w-0 before:border-2 before:opacity-0 hover:brightness-110 hover:before:w-full hover:before:opacity-100"
                  >
                    Login
                  </span>
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Header module with profile icon if logged in 1`] = `
<div>
  <div
    class="z-50 fixed duration-300 ease-in-out top-0  p-2  dark:bg-night-black-300 px-0 py-1 w-screen duration-300 ease-in-out dark min-h-[56px] translate-y-0"
  >
    <div
      class=" flex border-box justify-around w-full py-1 "
    >
      <div
        class="w-full max-w-7xl px-4 xl:px-0"
      >
        <div
          class=" grid  gap-4  p-2 grid-cols-2"
        >
          <a>
            children
          </a>
          <div
            class="md:hidden cursor-pointer flex flex-col max-w-min justify-center justify-self-end"
          >
            <div
              class="animate-hamburgertochevronleftreverse  border-night-black dark:border-max-white border-2 w-6"
            />
            <div
              class="animate-hamburgertochevronrightreverse  border-night-black dark:border-max-white border-2 w-6 mt-2"
            />
          </div>
          <nav
            class="col-span-2 md:col-span-1 duration-300 ease-in-out hidden h-0 md:flex md:h-full md:flex items-center justify-end"
          >
            <ul
              class="flex flex-col md:flex-row items-center justify-end"
            >
              <li
                class="inline mt-8 ml-12 md:mt-0"
              >
                <a>
                  children
                </a>
              </li>
              <li
                class="inline mt-8 ml-12 md:mt-0"
              >
                <a>
                  children
                </a>
              </li>
              <li
                class="inline mt-8 ml-12 md:mt-0"
              >
                <div
                  class="flex flex-col items-center justify-center w-full h-full"
                >
                  <div
                    class="cursor-pointer relative w-12 h-12"
                    data-testid="toggleProfile"
                  >
                    <div
                      class="absolute w-full h-full rounded-full bg-gradient-to-br from-blue-400 to-blue-600"
                    />
                  </div>
                </div>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</div>
`;
